Displayed datalist dropdown is out of sync with datalist options elements after DOM update

rwlbuis opened the pull request:
#### ba02cc6aade5caa1ddecb92778f6bfe82d5bef25
<pre>
Displayed datalist dropdown is out of sync with datalist options elements after DOM update
<a href="https://bugs.webkit.org/show_bug.cgi?id=201121">https://bugs.webkit.org/show_bug.cgi?id=201121</a>

Reviewed by Aditya Keerthi.

Adding datalist dropdown items by dynamically inserting options to the datalist on keydown/input change
was not taking effect right away. To fix this, detect datalist DOM changes through HTMLDataListElement::childrenChanged
and notify the corresponding input. This will eventually land in TextFieldInputType::dataListMayHaveChanged which
now ends with trying to display the suggestions if the input is focused.

Note that here a new DataListSuggestionActivationType value DataListMayHaveChanged is added to distinguish from
current usages (text changed, indicator/control clicked).

* LayoutTests/fast/forms/datalist/datalist-textinput-dynamically-add-options-expected.txt: Added.
* LayoutTests/fast/forms/datalist/datalist-textinput-dynamically-add-options-on-keydown-expected.txt: Added.
* LayoutTests/fast/forms/datalist/datalist-textinput-dynamically-add-options-on-keydown.html: Added.
* LayoutTests/fast/forms/datalist/datalist-textinput-dynamically-add-options.html: Added.
* Source/WebCore/html/DataListSuggestionInformation.h:
* Source/WebCore/html/HTMLDataListElement.cpp:
(WebCore::HTMLDataListElement::childrenChanged):
* Source/WebCore/html/HTMLDataListElement.h:
* Source/WebCore/html/TextFieldInputType.cpp:
(WebCore::TextFieldInputType::dataListMayHaveChanged):
(WebCore::TextFieldInputType::displaySuggestions):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/ios/WebDataListSuggestionsDropdownIOS.mm:
(-[WKDataListSuggestionsDropdown _displayWithActivationType:]):

Canonical link: <a href="https://commits.webkit.org/274608@main">https://commits.webkit.org/274608@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/92239574f9173faad8d87fe9caf3693d777f471d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/38897 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/17828 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/41242 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/41428 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/34614 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/41204 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/20678 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/15177 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/32597 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/39470 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/14998 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/33725 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13033 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/12990 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/34646 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/42705 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/35295 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/34986 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/38834 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/13707 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/11301 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/37047 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/15314 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/33935 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8858 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/14977 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/14801 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/html/TextFieldInputType.cpp:
```diff
@@ -911,6 +911,8 @@ void TextFieldInputType::dataListMayHaveChanged()
         return;
     m_dataListDropdownIndicator->setInlineStyleProperty(CSSPropertyDisplay, element()->list() ? CSSValueBlock : CSSValueNone, true);
+    if (element()->list() && visibleValue().length())
+        displaySuggestions(DataListSuggestionActivationType::DataListMayHaveChanged);
```
The `<datalist>` contents could change even when the input is empty, so I'm not sure why the `visibleValue` check is necessary?

rwlbuis replied:
You are right, checking for the input being focused is better, that is in the new version. The reason I was doing this kind of check is tested by datalist-textinput-dynamically-add-options.html, i.e. don't just show the picker with the new suggestions when the user is not yet interacting with the input only because datalist content was changed.

pxlcoder reviewed (commented):
Any chance you could upload a video of updating option elements on macOS and iOS? I would like to ensure the update looks reasonable to the user.

rwlbuis committed:
Displayed datalist dropdown is out of sync with datalist options elements after DOM update
https://bugs.webkit.org/show_bug.cgi?id=201121

Reviewed by Aditya Keerthi.

Adding datalist dropdown items by dynamically inserting options to the datalist on keydown/input change
was not taking effect right away. To fix this, detect datalist DOM changes through HTMLDataListElement::childrenChanged
and notify the corresponding input. This will eventually land in TextFieldInputType::dataListMayHaveChanged which
now ends with trying to display the suggestions if the input is focused.

Note that here a new DataListSuggestionActivationType value DataListMayHaveChanged is added to distinguish from
current usages (text changed, indicator/control clicked).

* LayoutTests/fast/forms/datalist/datalist-textinput-dynamically-add-options-expected.txt: Added.
* LayoutTests/fast/forms/datalist/datalist-textinput-dynamically-add-options-on-keydown-expected.txt: Added.
* LayoutTests/fast/forms/datalist/datalist-textinput-dynamically-add-options-on-keydown.html: Added.
* LayoutTests/fast/forms/datalist/datalist-textinput-dynamically-add-options.html: Added.
* Source/WebCore/html/DataListSuggestionInformation.h:
* Source/WebCore/html/HTMLDataListElement.cpp:
(WebCore::HTMLDataListElement::childrenChanged):
* Source/WebCore/html/HTMLDataListElement.h:
* Source/WebCore/html/TextFieldInputType.cpp:
(WebCore::TextFieldInputType::dataListMayHaveChanged):
(WebCore::TextFieldInputType::displaySuggestions):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/ios/WebDataListSuggestionsDropdownIOS.mm:
(-[WKDataListSuggestionsDropdown _displayWithActivationType:]):

Canonical link: https://commits.webkit.org/274608@main
